### PR TITLE
Feature/bot framework4 support

### PR DESCRIPTION
**Pull request by mpustelnik**

### CODE REVIEW

**Comment by jkolonko** [comment]
This solution uses different versions of Nuget packages, it would be nice to have them consolidated before releasing new library. Reference to Microsoft.NET.Test.SDK should not be necessary if XUnit is used.

File: src/Common/Objectivity.Bot.Tests.Stories/StoryModel/PerformanceStep.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
+﻿namespace Objectivity.Bot.Tests.Stories.StoryModel
+{
+    public class PerformanceStep<T> : IStoryStep<T>
+    {
+        private const string OptionsToken = "buttons";
```
Comment (piotrzajac) [inline src/Common/Objectivity.Bot.Tests.Stories/StoryModel/PerformanceStep.cs]: Consider making consts protected and removing from derived types.

File: src/Common/Objectivity.Bot.Tests.Stories/Exceptions/MissingOptionsException.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
+﻿namespace Objectivity.Bot.Tests.Stories.Exceptions
+{
+    using System;
+
+    public class MissingOptionsException : Exception
```
Comment (piotrzajac) [inline src/Common/Objectivity.Bot.Tests.Stories/Exceptions/MissingOptionsException.cs]: Consider following [best practices for exceptions](https://docs.microsoft.com/en-us/dotnet/standard/exceptions/best-practices-for-exceptions).

File: src/Common/Objectivity.Bot.Tests.Stories/Dialogs/WrappedDialogResult.cs [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+    [Serializable]
+    public class WrappedDialogResult
+    {
+        public WrappedDialogResult()
+        {
+            this.DialogStatus = DialogStatus.Idle;
+        }
+
+        public object Result { get; set; }
```
Comment (piotrzajac) [inline src/Common/Objectivity.Bot.Tests.Stories/Dialogs/WrappedDialogResult.cs]: Is it possible to use generic type with result?

File: src/Common/Objectivity.Bot.Tests.Stories/Dialogs/DialogStoryFrame.cs [hunk 4]
```diff
@@ -0,0 +1,41 @@
+﻿namespace Objectivity.Bot.Tests.Stories.Dialogs
+{
+    using System;
+    using Core;
+    using Newtonsoft.Json.Linq;
+    using StoryModel;
+
+    public class DialogStoryFrame<T> : IStoryFrame<T>
+    {
+        public DialogStoryFrame(DialogStatus dialogStatus,
```
[the remark this answers is not in the document]
Comment (mpustelnik) [reply src/Common/Objectivity.Bot.Tests.Stories/Dialogs/DialogStoryFrame.cs]: Treat warning as errors enabled. Above issue has been fixed.

File: src/V4/Objectivity.Bot.Tests.Stories.Xunit.V4.DemoBot/Dialogs/TestGreetingDialog.cs [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+            userState.UserName = name;
+            
+            await this.userStateAccessor.UserState.SaveChangesAsync(stepContext.Context, cancellationToken: cancellationToken);
+
+            return await stepContext.PromptAsync(
+                AgePrompt,
+                new PromptOptions
+                {
+                    Prompt = MessageFactory.Text($"How many floors have you climbed today?")
```
[the remark this answers is not in the document]
Comment (mpustelnik) [reply src/V4/Objectivity.Bot.Tests.Stories.Xunit.V4.DemoBot/Dialogs/TestGreetingDialog.cs]: Test scenario changed

File: src/Common/Objectivity.Bot.Tests.Stories/Player/IStoryResult.cs [hunk 6]
```diff
@@ -0,0 +1,9 @@
+﻿namespace Objectivity.Bot.Tests.Stories.Player
+{
+    using System.Collections.Generic;
+
+    public interface IStoryResult
+    {
+        IDictionary<string, object> OutputValues { get; }
```
[the remark this answers is not in the document]
Comment (mpustelnik) [reply src/Common/Objectivity.Bot.Tests.Stories/Player/IStoryResult.cs]: IStoryResult has been removed as it was redundant.

File: src/Common/Objectivity.Bot.Tests.Stories/Dialogs/WrappedDialogResult.cs (hunk 3 above)
Comment (mpustelnik) [reply src/Common/Objectivity.Bot.Tests.Stories/Dialogs/WrappedDialogResult.cs]: That would impact framework usage complexity. At the moment result type is given only when its necessary eg. DialogDoneWithResult<int>(x => x == 17)

File: src/Common/Objectivity.Bot.Tests.Stories/StoryModel/PerformanceStep.cs (hunk 1 above)
Comment (mpustelnik) [reply src/Common/Objectivity.Bot.Tests.Stories/StoryModel/PerformanceStep.cs]: Fixed.

File: src/Common/Objectivity.Bot.Tests.Stories/Exceptions/MissingOptionsException.cs (hunk 2 above)
Comment (mpustelnik) [reply src/Common/Objectivity.Bot.Tests.Stories/Exceptions/MissingOptionsException.cs]: Suggested constructor overloads have been added.

**Pull request merged**